grant-project/zcash-grant-system
# Contribution Admin

Closes #135.

### What This Does

Provides a contribution admin page where admins can do the following:

* List, search, filter, and sort contributions
* Edit existing contributions
* Create new contributions
* Removes a lot of the `Decimal.normalize` calls I had before, realized they didn't do the right thing. This will allow for trailing zeroes in the database. If we're worried about this, we can do a little regex replace before inserting into db to keep them consistent.

In the process of doing this, I reworked a decent bit of the pagination code to make it flexible in the admin, cc @AMStrix. Should be good for converting users to it too.

### Steps to Test

* Create a contribution as a user
* Go into the admin and find the contribution
* Edit the contribution and give it a txid (any will do)
* Confirm it's now got a CONFIRMED status
* Create a new contribution without a txid, confirm it has a PENDING status
* Try to put garbage in the form (except txid, it does no real validation) and confirm it errors

## Screenshots

<img width="1247" alt="screen shot 2019-02-06 at 5 35 42 pm" src="https://user-images.githubusercontent.com/649992/52379176-b7feae80-2a37-11e9-8d66-3158afb879d5.png">


<img width="1247" alt="screen shot 2019-02-06 at 5 35 46 pm" src="https://user-images.githubusercontent.com/649992/52379179-baf99f00-2a37-11e9-8b18-40ee3c00a290.png">
